sycl-fusion/jit-compiler/lib/KernelFusion.cpp

```diff
@@ -0,0 +1,116 @@
+  // information. We could remove the copy, if we removed the const from the
+  // input interface, so it depends on the guarantees we want to give to
+  // callers.
+  ModuleInfo.kernels().insert(ModuleInfo.kernels().end(),
+                              KernelInformation.begin(),
+                              KernelInformation.end());
+  // Load all input kernels from their respective SPIR-V modules into a single
+  // LLVM IR module.
+  auto ModOrError = translation::SPIRVLLVMTranslator::loadSPIRVKernels(
+      *JITCtx.getLLVMContext(), ModuleInfo.kernels());
+  if (auto Error = ModOrError.takeError()) {
+    return errorToFusionResult(std::move(Error), "SPIR-V translation failed");
+  }
+  auto LLVMMod = std::move(*ModOrError);
+
+  // Add information about the kernel that should be fused as metadata into the
+  // LLVM module.
+  FusedFunction FusedKernel{FusedKernelName, KernelsToFuse,
+                            std::move(Identities), Internalization, Constants};
+  FusedFunctionList FusedKernelList;
+  FusedKernelList.push_back(FusedKernel);
+  auto NewModOrError =
+      helper::FusionHelper::addFusedKernel(LLVMMod.get(), FusedKernelList);
+  if (auto Error = NewModOrError.takeError()) {
+    return errorToFusionResult(std::move(Error),
+                               "Insertion of fused kernel stub failed");
+  }
+  auto NewMod = std::move(*NewModOrError);
+
+  // TODO: Invoke the actual fusion via LLVM pass manager.
+  // This will be added in a later PR.
+  auto NewModInfo = std::make_unique<SYCLModuleInfo>();
+
+  // Get the updated kernel info for the fused kernel and add the information to
+  // the existing KernelInfo.
+  if (!NewModInfo->hasKernelFor(FusedKernelName)) {
+    return FusionResult{"No KernelInfo for fused kernel"};
+  }
+
+  auto &FusedKernelInfo = *NewModInfo->getKernelFor(FusedKernelName);
+
+  // Translate the LLVM IR module resulting from the fusion pass into SPIR-V.
+  auto BinaryOrError =
+      translation::SPIRVLLVMTranslator::translateLLVMtoSPIRV(*NewMod, JITCtx);
+  if (auto Error = BinaryOrError.takeError()) {
+    return errorToFusionResult(std::move(Error),
+                               "Translation to SPIR-V failed");
+  }
+  auto *SPIRVBin = *BinaryOrError;
+
+  // Update the KernelInfo for the fused kernel with the address and size of the
+  // SPIR-V binary resulting from translation.
+  auto &FusedBinaryInfo = FusedKernelInfo.BinaryInfo;
+  FusedBinaryInfo.Format = BinaryFormat::SPIRV;
+  // Output SPIR-V should use the same number of address bits as the input
+  // SPIR-V. SPIR-V translation requires all modules to use the same number of
+  // address bits, so it's safe to take the value from the first one.
+  FusedBinaryInfo.AddressBits =
+      ModuleInfo.kernels().front().BinaryInfo.AddressBits;
+  FusedBinaryInfo.BinaryStart = SPIRVBin->address();
+  FusedBinaryInfo.BinarySize = SPIRVBin->size();
+
+  return FusionResult{FusedKernelInfo};
+}
```

Comment:
This function uses `auto` a lot and though the reader may have a vague idea of what some of these are, most of them wouldn't be all that more verbose with explicitly stated types and then people would know for sure what they are when they read this.

Reply:
I reduced the use of `auto` in this file/function and also other implementation files.